lalrpop/src/lr1/lane_table/construct/group.rs

```diff
@@ -0,0 +1,165 @@
+use lr1::lane_table::table::context_set::{ContextSet};
+use lr1::core::*;
+use std::cmp::max;
+use std::fmt;
+use std::mem::replace;
+use ena::unify::{UnifyKey, UnifyValue, UnificationTable};
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Group {
+    index: usize
+}
+
+impl fmt::Debug for Group {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "G{}", self.index)
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct GroupUnifyKey {
+    index : u32
+}
+
+impl UnifyKey for GroupUnifyKey {
+    type Value = Group;
+
+    fn index(&self) -> u32 {
+        self.index
+    }
+
+    fn from_index(u: u32) -> Self {
+        GroupUnifyKey { index: u }
+    }
+
+    fn tag() -> &'static str {
+        "GroupUnifyKey"
+    }
+}
+
+impl UnifyValue for Group {
+    fn unify_values(value1: &Self, value2: &Self) -> Result<Self, (Self, Self)> {
+        // correctness of merging algorithm relies on order used here
+        Ok(max(*value1, *value2))
+    }
+}
+
+pub struct Groups {
+    ///! groups maps state index into group
```

Comment:
Nit: `///`, not `///!` (`//!` attaches a comment the enclosing item, but that's not what you want here)